src/persist-types/src/arrow.rs

```diff
@@ -622,39 +624,36 @@ impl ArrayBound {
 }
 
 /// Makes a best effort to shrink the proto while preserving the ordering.
-/// (The proto may not be smaller after this method is called, but it should always
+/// (The proto might not be smaller after this method is called, but it should always
 /// be a valid lower bound.)
-fn maybe_trim_proto(proto: &mut ProtoArrayData, max_len: usize) {
+fn maybe_trim_proto(data_type: &mut proto::DataType, body: &mut ProtoArrayData, max_len: usize) {
+    assert!(body.data_type.is_none(), "expected separate data type");
     // TODO: consider adding cases for strings and byte arrays
-    let encoded_len = proto.encoded_len();
-    match &mut proto.data_type {
-        Some(proto::DataType {
-            kind:
-                Some(proto::data_type::Kind::Struct(proto::data_type::Struct { children: fields })),
-        }) => {
+    let encoded_len = data_type.encoded_len() + body.encoded_len();
+    match &mut data_type.kind {
+        Some(data_type::Kind::Struct(data_type::Struct { children: fields })) => {
             // Pop off fields one by one, keeping an estimate of the encoded length.
             let mut struct_len = encoded_len;
             while struct_len > max_len {
-                let Some(mut child) = proto.children.pop() else {
+                let Some(mut child) = body.children.pop() else {
                     break;
                 };
                 let Some(mut field) = fields.pop() else { break };
```

Comment:
nit: Not part of this PR so feel free to ignore, but it seems like it would be nice to at least have a `soft_assert!` here that `child` and `field` have the same data type